src/dotnet-watch/Program.cs

```diff
@@ -121,7 +125,19 @@ private void OnCancelKeyPress(object sender, ConsoleCancelEventArgs args)
                 _reporter.Output("Shutdown requested. Press Ctrl+C again to force exit.");
             }
 
-            _cts.Cancel();
+            // always fire user interupt inline because this used to determine if the process
+            // is exiting due to user interaction or not
+            _userInterrupt.Cancel();
+
+            if (!Thread.CurrentThread.IsThreadPoolThread)
+            {
+                // Invoke the cancellation on the default thread pool to workaround https://github.com/dotnet/corefx/issues/29699
+                ThreadPool.QueueUserWorkItem(_ => _cts.Cancel());
+            }
+            else
+            {
+                _cts.Cancel();
```

Comment:
Is there a scenario where QUWI doesn't work? I'd always cancel the CTS using QUWI for consistency.

Reply:
I was getting a race condition in the unit tests that fakes the CTRL+C behavior.

Reply:
No big deal either way. I'd probably create a helper method that verifies either that the token is already canceled or becomes canceled in 30 seconds.

Reply:
actually, tested this more. The race condition was a different problem. Updated to always dispatch now.